fat32 fs creates case-sensitive dirs and files

The fat32 filesystem implementation allows creating directories and files that are distinct on case (upper/lower). I can for example have all these created (as listed by `7z l file`):
```
...
2024-07-19 10:45:28 .....            5          512  EFI/BOOT/BOOTX64.EFI
2024-07-19 10:45:28 .....           17          512  EFI/BOOT/bootx64.efi
2024-07-19 10:45:28 .....           17          512  EFI/boot/bootx64.efi
...
```

Hello, I encountered the same issue (with /EFI/BOOT as well, lol).

I was able to reproduce the duplicate folders with a test case added to fat32_test.go:

```go
func TestFat32MkdirCases(t *testing.T) {
	f, err := tmpFat32(false, 0, 0)
	if err != nil {
		t.Fatal(err)
	}
	defer os.Remove(f.Name())
	fs, err := fat32.Create(f, 1048576, 0, 512, "")
	if err != nil {
		t.Error(err.Error())
	}
	err = fs.Mkdir("/EFI/BOOT")
	if err != nil {
		t.Error(err.Error())
	}
	// Make the same folders but now lowercase ... I expect it not to create anything new,
	// these folders exist but are named /EFI/BOOT
	err = fs.Mkdir("/efi/boot")
	if err != nil {
		t.Error(err.Error())
	}
	files, err := fs.ReadDir("/")
	if err != nil {
		t.Error(err.Error())
	}
	if len(files) != 1 {
		for _, file := range files {
			fmt.Printf("file: %s\n", file.Name())
		}
		t.Fatalf("expected 1 file, found %d", len(files))
	}
}
```
Test result:
```
=== RUN   TestFat32MkdirCases
file: EFI
file: efi
    fat32_test.go:136: expected 1 file, found 2
--- FAIL: TestFat32MkdirCases (0.00s)

FAIL
```

--- a/filesystem/fat32/fat32.go
+++ b/filesystem/fat32/fat32.go
@@ -830,7 +830,11 @@ func (fs *FileSystem) readDirWithMkdir(p string, doMake bool) (*Directory, []*di
 		// do we have an entry whose name is the same as this name?
 		found := false
 		for _, e := range entries {
-			if e.filenameLong != subp && e.filenameShort != subp && (!e.lowercaseShortname || (e.lowercaseShortname && !strings.EqualFold(e.filenameShort, subp))) {
+			// if the filename does not match, continue
+			// match is determined by any one of:
+			// - long filename == provided name
+			// - uppercase(short filename) == uppercase(provided name)
+			if e.filenameLong != subp && !strings.EqualFold(e.filenameShort, subp) {
 				continue
 			}
 			if !e.isSubdirectory {

--- a/filesystem/fat32/fat32_test.go
+++ b/filesystem/fat32/fat32_test.go
@@ -946,3 +946,35 @@ func TestFat32Label(t *testing.T) {
 		}
 	})
 }
+
+func TestFat32MkdirCases(t *testing.T) {
+	f, err := tmpFat32(false, 0, 0)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer os.Remove(f.Name())
+	fs, err := fat32.Create(f, 1048576, 0, 512, "")
+	if err != nil {
+		t.Error(err.Error())
+	}
+	err = fs.Mkdir("/EFI/BOOT")
+	if err != nil {
+		t.Error(err.Error())
+	}
+	// Make the same folders but now lowercase ... I expect it not to create anything new,
+	// these folders exist but are named /EFI/BOOT
+	err = fs.Mkdir("/efi/boot")
+	if err != nil {
+		t.Error(err.Error())
+	}
+	files, err := fs.ReadDir("/")
+	if err != nil {
+		t.Error(err.Error())
+	}
+	if len(files) != 1 {
+		for _, file := range files {
+			fmt.Printf("file: %s\n", file.Name())
+		}
+		t.Fatalf("expected 1 file, found %d", len(files))
+	}
+}

